Source/core/SocketPort.cpp

```diff
@@ -373,7 +373,12 @@ namespace Core {
 
         // Make sure the socket is closed before you destruct. Otherwise
         // the virtuals might be called, which are destructed at this point !!!!
-        ASSERT(m_Socket == INVALID_SOCKET);
+        ASSERT((m_Socket == INVALID_SOCKET) &&(m_State == 0));
```

Comment:
Please see the comment above the ASSERT. Checking for m_State == 0 is not usefull as you already check that the m_Socket should be invalid and thus not usable, and thus m_state should not be checked/used.
Unregistering at this point is even dangerous and might lead to race conditions. If the socket is still open and the destructor is called, the VTABLE might be in a destructed mode and have dangling pointers (hence the ASSERT check).
This can be properly solved by making sure that the class inheriting and implementing the virtuals calls Close(Core::Infinite); in its destructor. This way the full close cycle is executed (and thus all implementations from the virtuals are called to signal a closure) before they are destructed.

Reply:
but these changes are already available in master branch.